pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -136,35 +154,63 @@ func (sf *IteratorDatastore) ReadStartingWithUser(
 	if found {
 		keyItem.counter++
 		sf.internalStorage.mu.Unlock()
+		keyItem.iter.mu.Lock()
+
 		span.SetAttributes(attribute.Bool("found", true))
 
-		// For now, this will be identical to the not found path.  When the SF iterator datastore
-		// is fully implemented, the found path will look up the shared iterator and create
-		// a clone from it.
-	} else {
-		if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
-			sf.internalStorage.mu.Unlock()
-			sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadStartingWithUser).Inc()
-			// we cannot share this iterator because we have reached the size limit.
+		// by the time we have access to keyItem.iter.mu, we know that
+		// the inner is already set OR the queryErr is set (for which clone
+		// will fail).
+		item, err := keyItem.iter.clone()
+		if err != nil {
+			keyItem.iter.mu.Unlock()
```

Comment:
This can be moved on stop of this `if` block and used with a `defer`.

Reply:
actually, I want this to unlock before calling `return sf.RelationshipTupleReader.ReadStartingWithUser()`.